File: hookwise/coaching/metacognition.py
"""Metacognition reminder engine for hookwise coaching.

Provides time-gated metacognitive prompts that encourage the user
to pause and reflect during Claude Code sessions. Prompts are shown
at a configurable interval (default: 5 minutes) and avoid repeating
recent prompts.

Prompt selection weights toward mode-relevant prompts when available
and draws from a configurable prompt pool (built-in defaults or
external JSON file).
"""

from __future__ import annotations

import json
import logging
import random
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

logger = logging.getLogger("hookwise")

# Default interval between metacognition prompts (in seconds)
DEFAULT_INTERVAL_SECONDS = 300  # 5 minutes

# Default metacognition prompts (built into code)
DEFAULT_METACOGNITION_PROMPTS: list[dict[str, str]] = [
    {"id": "meta_01", "text": "Pause: Are you solving the right problem?"},
    {"id": "meta_02", "text": "One idea. One pause. Let them ask."},
    {"id": "meta_03", "text": "What would you explain to a junior dev about this change?"},
]

# Rapid acceptance prompts
DEFAULT_RAPID_ACCEPTANCE_PROMPTS: list[dict[str, str]] = [
    {"id": "rapid_01", "text": "You just accepted a large AI change very quickly. Can you explain what it does?"},
]


def load_prompts_file(prompts_file: str | None) -> dict[str, Any] | None:
    """Load prompts from an external JSON file.

    The file should contain a JSON object with keys like
    "metacognition", "builder_trap", "rapid_acceptance", each
    containing a list of prompt dicts with "id" and "text".

    Args:
        prompts_file: Path to the JSON prompts file (may contain ~).
            If None, returns None.

    Returns:
        Parsed dict from the file, or None if the file doesn't exist
        or is malformed.
    """
    if not prompts_file:
        return None

    try:
        path = Path(prompts_file).expanduser()
        if not path.is_file():
            return None
        content = path.read_text(encoding="utf-8")
        data = json.loads(content)
        if isinstance(data, dict):
            return data
        return None
    except (OSError, json.JSONDecodeError, ValueError) as exc:
        logger.debug("Could not load prompts file %s: %s", prompts_file, exc)
        return None


def check_metacognition_interval(
    cache: dict[str, Any],
    now: datetime,
    config: dict[str, Any],
) -> bool:
    """Check whether enough time has passed for a metacognition prompt.

    Args:
        cache: The coaching cache dict.
        now: Current timestamp.
        config: The metacognition config section.

    Returns:
        True if the interval has elapsed, False otherwise.
    """
    if not config.get("enabled", True):
        return False

    interval = config.get("interval_seconds", DEFAULT_INTERVAL_SECONDS)

    last_prompt_at = cache.get("last_prompt_at")
    if not last_prompt_at:
        # No prompt has been shown yet -- show one
        return True

    try:
        last_dt = datetime.fromisoformat(last_prompt_at)
        elapsed = (now - last_dt).total_seconds()
        return elapsed >= interval
    except (ValueError, TypeError):
        # Malformed timestamp -- show prompt
        return True


def select_metacognition_prompt(
    cache: dict[str, Any],
    current_mode: str,
    config: dict[str, Any],
) -> dict[str, str] | None:
    """Select a metacognition prompt, avoiding recent repeats.

    Weights toward mode-relevant prompts when available in the pool.
    Falls back to any available prompt if no mode-specific ones exist.

    Args:
        cache: The coaching cache dict.
        current_mode: The current classified mode.
        config: The metacognition config section.

    Returns:
        Dict with "id" and "text", or None if no prompts available.
    """
    # Load prompts from file or use defaults
    prompts_file = config.get("prompts_file")
    external_prompts = load_prompts_file(prompts_file)

    if external_prompts and "metacognition" in external_prompts:
        pool = external_prompts["metacognition"]
    else:
        pool = DEFAULT_METACOGNITION_PROMPTS

    if not pool:
        return None

    # Get recent prompt IDs to avoid
    prompt_history = cache.get("prompt_history", [])
    recent = set(prompt_history[-3:]) if prompt_history else set()

    # Filter out recently shown prompts
    candidates = [p for p in pool if p.get("id") not in recent]

    # If all filtered out, use full pool
    if not candidates:
        candidates = list(pool)

    # Weight toward mode-relevant prompts if they have a "modes" field
    mode_relevant = [
        p for p in candidates
        if current_mode in p.get("modes", [])
    ]

    if mode_relevant:
        return random.choice(mode_relevant)

    return random.choice(candidates)

File: src/core/coaching/metacognition.ts
/**
 * Metacognition Reminder Engine for hookwise coaching system.
 *
 * Checks elapsed time, detects behavioral triggers, and selects
 * contextually relevant prompts while avoiding repetition.
 *
 * All functions are synchronous and fail-open.
 */

import { readFileSync, existsSync } from "node:fs";
import type {
  HookPayload,
  CoachingConfig,
  CoachingCache,
  MetacognitionResult,
  AlertLevel,
} from "../types.js";
import { classifyMode, computeAlertLevel } from "./builder-trap.js";

/** Minimum lines for rapid acceptance detection. */
const RAPID_ACCEPTANCE_MIN_LINES = 50;

/** Maximum seconds for rapid acceptance detection. */
const RAPID_ACCEPTANCE_MAX_SECONDS = 5;

/** Number of recent prompts to avoid repeating. */
const HISTORY_WINDOW = 3;

/**
 * Built-in metacognition prompts organized by category.
 */
interface PromptEntry {
  id: string;
  text: string;
  category: string;
}

const DEFAULT_PROMPTS: PromptEntry[] = [
  { id: "meta-1", text: "What assumption am I making right now that I haven't verified?", category: "reflection" },
  { id: "meta-2", text: "Am I solving the right problem, or the problem I want to solve?", category: "reflection" },
  { id: "meta-3", text: "If I had to explain this approach to a colleague, would it make sense?", category: "clarity" },
  { id: "meta-4", text: "What's the simplest thing that could possibly work here?", category: "simplicity" },
  { id: "meta-5", text: "Am I building what was asked for, or what I think should be built?", category: "alignment" },
  { id: "meta-6", text: "Have I tested my understanding before writing more code?", category: "verification" },
  { id: "meta-7", text: "Is this complexity essential, or am I over-engineering?", category: "simplicity" },
  { id: "meta-8", text: "When was the last time I stepped back and reviewed the big picture?", category: "reflection" },
  { id: "meta-9", text: "Am I learning something new, or just going through the motions?", category: "growth" },
  { id: "meta-10", text: "What would I do differently if I had to start over right now?", category: "reflection" },
  { id: "meta-11", text: "One idea. One pause. Let them ask.", category: "communication" },
  { id: "meta-12", text: "Trust the pull. Don't push.", category: "communication" },
];

/**
 * Load custom prompts from a JSON file, falling back to empty array.
 */
function loadCustomPrompts(filePath: string): PromptEntry[] {
  try {
    if (!existsSync(filePath)) return [];
    const content = readFileSync(filePath, "utf-8");
    const parsed = JSON.parse(content);
    if (!Array.isArray(parsed)) return [];
    return parsed.filter(
      (p: unknown) =>
        typeof p === "object" &&
        p !== null &&
        "id" in p &&
        "text" in p &&
        "category" in p
    ) as PromptEntry[];
  } catch {
    return [];
  }
}

/**
 * Select a prompt that avoids the last N shown.
 * Prefers prompts matching the current mode category.
 */
function selectPrompt(
  prompts: PromptEntry[],
  recentIds: string[],
  _currentMode: string
): PromptEntry | null {
  const recentSet = new Set(recentIds.slice(-HISTORY_WINDOW));
  const eligible = prompts.filter((p) => !recentSet.has(p.id));

  if (eligible.length === 0) {
    // All prompts exhausted; pick any
    if (prompts.length === 0) return null;
    return prompts[Math.floor(Math.random() * prompts.length)];
  }

  // Randomize within eligible
  return eligible[Math.floor(Math.random() * eligible.length)];
}

/**
 * Determine the most important trigger type for this check.
 *
 * Priority: rapid_acceptance > builder_trap > mode_change > interval
 */
function determineTriggerType(
  cache: CoachingCache,
  config: CoachingConfig,
  event: HookPayload,
  intervalElapsed: boolean
): MetacognitionResult["triggerType"] | null {
  // 1. Rapid acceptance: > 50 lines in < 5s
  if (cache.lastLargeChange) {
    if (
      cache.lastLargeChange.linesChanged > RAPID_ACCEPTANCE_MIN_LINES &&
      cache.lastLargeChange.acceptedWithinSeconds < RAPID_ACCEPTANCE_MAX_SECONDS
    ) {
      return "rapid_acceptance";
    }
  }

  // 2. Builder trap escalation: alert level went up
  const currentAlertLevel = computeAlertLevel(cache, config.builderTrap.thresholds);
  if (isEscalation(cache.alertLevel, currentAlertLevel)) {
    return "builder_trap";
  }

  // 3. Mode change: tool classifies to a different mode than cached
  if (event.tool_name) {
    const eventMode = classifyMode(
      event.tool_name,
      (event.tool_input ?? {}) as Record<string, unknown>,
      config.builderTrap
    );
    if (eventMode !== cache.currentMode) {
      return "mode_change";
    }
  }

  // 4. Interval
  if (intervalElapsed) {
    return "interval";
  }

  return null;
}

/** Alert level ordering for escalation detection. */
const ALERT_ORDER: Record<AlertLevel, number> = {
  none: 0,
  yellow: 1,
  orange: 2,
  red: 3,
};

function isEscalation(previous: AlertLevel, current: AlertLevel): boolean {
  return ALERT_ORDER[current] > ALERT_ORDER[previous];
}

/**
 * Check whether a metacognition prompt should be emitted.
 *
 * @param event - The current hook payload
 * @param config - Full coaching configuration
 * @param cache - Mutable coaching cache (updates cache.alertLevel, lastPromptAt, and promptHistory)
 * @param now - Current timestamp ISO string (for testability)
 * @returns MetacognitionResult with shouldEmit and optional prompt details
 */
export function checkAndEmit(
  event: HookPayload,
  config: CoachingConfig,
  cache: CoachingCache,
  now: string
): MetacognitionResult {
  // Disabled check
  if (!config.metacognition.enabled) {
    return { shouldEmit: false };
  }

  const nowMs = new Date(now).getTime();
  const lastPromptMs = new Date(cache.lastPromptAt).getTime();
  const elapsedSeconds = (nowMs - lastPromptMs) / 1000;
  const intervalElapsed = elapsedSeconds >= config.metacognition.intervalSeconds;

  // Determine trigger type
  const triggerType = determineTriggerType(cache, config, event, intervalElapsed);

  // Update alert level to prevent repeated escalation triggers
  const currentAlertLevel = computeAlertLevel(cache, config.builderTrap.thresholds);
  cache.alertLevel = currentAlertLevel;

  if (!triggerType) {
    return { shouldEmit: false };
  }

  // Load prompts
  let prompts = [...DEFAULT_PROMPTS];
  if (config.metacognition.promptsFile) {
    const custom = loadCustomPrompts(config.metacognition.promptsFile);
    prompts = [...prompts, ...custom];
  }

  // Select a non-repeating prompt
  const selected = selectPrompt(prompts, cache.promptHistory, cache.currentMode);
  if (!selected) {
    return { shouldEmit: false };
  }

  // Update cache tracking for prompt emission
  cache.lastPromptAt = now;
  cache.promptHistory = [...cache.promptHistory.slice(-(HISTORY_WINDOW - 1)), selected.id];

  return {
    shouldEmit: true,
    promptText: selected.text,
    promptId: selected.id,
    category: selected.category,
    triggerType,
  };
}
